decentraland/unity-renderer
# Fix: mesh combiner optimizations

## What does this PR change?

Closes #4134

1. Heavy optimizations of `AvatarMeshCombinerHelper` and downwards:
  - Main focus on allocations
  - Lesser focus on suboptimal paths and redundant logic execution
  - Skip several things that I consider to be out of the scope as they tend to become a rabbit hole

## Allocations Analysis

### Before

![image](https://user-images.githubusercontent.com/118179774/217291388-e33afde0-6483-43d6-b92b-187e173449c9.png)
![image](https://user-images.githubusercontent.com/118179774/217292206-4ef86e7c-daae-46dc-886b-90d5f46148c5.png)

### After

P.S. Deep profiling does not reflect CPU consumption

<img width="867" alt="image" src="https://user-images.githubusercontent.com/118179774/217277623-4b24886e-6ec8-41a0-a47d-d21527d87199.png">

![image](https://user-images.githubusercontent.com/118179774/217275182-8dc4e039-d5b3-4946-a6e9-e2e306da9c5e.png)
![image](https://user-images.githubusercontent.com/118179774/217277976-27a22ebc-cb10-4c25-8c98-746dfa70acd0.png)
![image](https://user-images.githubusercontent.com/118179774/217278708-954dfa81-7233-4960-84ce-c2b1589874b1.png)
![image](https://user-images.githubusercontent.com/118179774/217279879-7c4b0769-bcf4-42c3-9947-161395963f50.png)
![image](https://user-images.githubusercontent.com/118179774/217286386-ece65513-dbfc-4868-afb3-c017d62e7921.png)
![image](https://user-images.githubusercontent.com/118179774/217287913-7c77a276-d742-4e24-85cc-a20c8967f0e5.png)

## CPU

### Before
![image](https://user-images.githubusercontent.com/118179774/217295248-c97124d0-7f48-4524-b077-18bc0d15ed49.png)
![image](https://user-images.githubusercontent.com/118179774/217295486-2ba8918e-60e1-451e-861f-b43c3033844b.png)
![image](https://user-images.githubusercontent.com/118179774/217295840-71207477-2ff2-4614-acd2-1c984223f778.png)

### After
<img width="989" alt="image" src="https://user-images.githubusercontent.com/118179774/217297561-0d24b638-5ea3-4153-aec6-2a5dbf5a99c0.png">
<img width="924" alt="image" src="https://user-images.githubusercontent.com/118179774/217297894-df5f6435-eedf-4f68-960e-3037bb34955f.png">
<img width="986" alt="image" src="https://user-images.githubusercontent.com/118179774/217298376-58e3bf48-7978-44fd-907b-c9528c6962ee.png">

Pretty much random CPU usage, I will leave it for now as I don't have more time for investigation


## How to test the changes?



1. Go to: https://play.decentraland.zone/?explorer-branch={branch_name}&{desired_url_params}
3. Ensure there are no visual glitches or other bugs connected to the avatars: they should work as fine as before

## Our Code Review Standards

https://github.com/decentraland/unity-renderer/blob/master/docs/code-review-standards.md
